capi/geos_ts_c.cpp

```diff
@@ -329,14 +330,76 @@ gstrdup(std::string const& str)
 
 } // namespace anonymous
 
+// Execute a lambda, using the given context handle to process errors.
+// Return errval on error.
+template<typename RT, typename F>
+inline RT execute(GEOSContextHandle_t extHandle, RT errval, F&& f) {
+    if (extHandle == nullptr) {
+        return errval;
+    }
+
+    GEOSContextHandleInternal_t* handle = reinterpret_cast<GEOSContextHandleInternal_t*>(extHandle);
+    if (!handle->initialized) {
+        return errval;
+    }
+
+    try {
+        return f();
+    } catch (const std::exception& e) {
+        handle->ERROR_MESSAGE("%s", e.what());
+    } catch (...) {
+        handle->ERROR_MESSAGE("Unknown exception thrown");
+    }
+
+    return errval;
+}
+
+// Execute a lambda, using the given context handle to process errors.
+// Return nullptr on error.
+template<typename F, typename std::enable_if<!std::is_void<decltype(std::declval<F>()())>::value, F>::type* = nullptr>
```

Comment:
Compile error in MSVC 2019, C++20 strict:
`D:\a\1\s\capi\geos_ts_c.cpp(359,84): error C2064:  term does not evaluate to a function taking 0 arguments [D:\a\1\s\build.cxx20\geos_c.vcxproj]`

Log: https://dev.azure.com/libgeos/geos/_build/results?buildId=64

Reply:
Possibly related to https://developercommunity.visualstudio.com/content/problem/726778/this-snippet-compiles-totally-fine-on-godbolt-msvc.html

Reply:
Follow-up shows they implemented a fix for this in MSVC https://developercommunity.visualstudio.com/comments/744771/view.html

Reply:
Thanks @hobu . It seems though that your suggestion for better error messages (https://trac.osgeo.org/geos/ticket/989) would tip the scales towards a macro-based solution instead?

Reply:
Your template solution is nice and clean. I think it should be merged. geos_ts_c.cpp is a mess of copypasta right now.